Make session storage injectable

SmellyFish opened the pull request:
These changes will make the session storage object injectable into the `LinkedIn` class. This will allow developers to write custom implementations of the `DataStorageInterface` interface and use them with this library without having to modify any source code. This will allow developers to bring this package into their project (and keep it updated) via composer while implementing the concrete session storage logic inside of their own project.

The code explicitly enforces the passed in `$storage` variable to be of type `DataStorageInterface`, so the rest of the package can safely remain oblivious of the concrete implementation and keep using it through the interface.

Nyholm commented:
Thank you for this PR. It has been lot more issues with this data storage then I could expect. I see that your solution really takes care of the problem. I'll be happy to merge it but there are some minor issues. 

See the following code comments.

Nyholm commented on .gitignore:
```diff
@@ -2,3 +2,4 @@ build
 phpunit.xml
 composer.lock
 vendor
+.idea
```
It is considered best practice to put this in your global git-ignore file. Please remove this.

Nyholm commented on src/Happyr/LinkedIn/LinkedIn.php:
```diff
@@ -90,31 +90,41 @@ class LinkedIn
+     * @param DataStorageInterface $storage [optional] the session storage object to use
      */
-    public function __construct($appId, $appSecret)
+    public function __construct($appId, $appSecret, $storage = null)
```
You should annotate the third parameter with DataStorageInterface

Nyholm commented on src/Happyr/LinkedIn/LinkedIn.php:
```diff
@@ -90,31 +90,41 @@ class LinkedIn
+     * @param DataStorageInterface $storage [optional] the session storage object to use
      */
-    protected function init()
+    protected function init($storage = null)
```
Annotate here as well

Nyholm commented on src/Happyr/LinkedIn/LinkedIn.php:
```diff
@@ -90,31 +90,41 @@ class LinkedIn
+            $this->storage = $storage;
         } else {
-            $this->storage = new SessionStorage();
+            // Use the Illuminate Session storage if it is available
```
I think we should just default to the `SessionStorage`. If you want to use the `IlluminateSessionStorage` you have to inject it.

Nyholm commented on tests/Happyr/LinkedIn/LinkedInTest.php:
```diff
@@ -35,6 +35,22 @@ public function testConstructor()
+        new LinkedIn(self::APP_ID, self::APP_SECRET, new \stdClass());
+    }
+    
+    public function testInitProperlyAssignsPassedInStorage()
```
Great test! Do also test that the default `SessionStorage` is used when we don't have a third parameter.

SmellyFish commented:
Thank you for taking a look at my PR. I just addressed all of your comments.

SmellyFish commented:
Oops, looks like I broke the build :crying_cat_face: I'm taking a look at it now.

Nyholm commented on tests/Happyr/LinkedIn/LinkedInTest.php:
```diff
@@ -35,6 +36,19 @@ public function testConstructor()
             'Expect the API secret to be set.');
     }
 
+    public function testInitProperlyAssignsPassedInStorage()
```
This test does not do much.. Try change the init function and you will see that this test still will pass. Consider changing it to: 

```php
        $sessionStorage = new IlluminateSessionStorage();
        $ln = new LinkedIn(self::APP_ID, self::APP_SECRET, $sessionStorage);
        $this->assertInstanceOf('Happyr\LinkedIn\Storage\IlluminateSessionStorage', $ln->getStorage());
```

SmellyFish replied:
Good call. I think using an `IlluminateSessionStorage` object is a good measure. But I think leaving the assertion as `$this->assertEquals($sessionStorage, $ln->getStorage());` is more prudent as it will make sure that the actual variable we passed to the constructor got assigned to the member variable. `$this->assertInstanceOf(...)` only guarantees that the member variable is of the same instance as the passed in variable, not that the variable was retained properly by the object. What do you think?

Nyholm replied:
Hm, you are correct but you have one flaw in your statement. 
> $this->assertEquals($sessionStorage, $ln->getStorage()); is more prudent as it will make sure that the actual variable we passed to the constructor got assigned to the member variable.

This is not true. See the following example: 
```php 
// IlluminateSessionStorage extends SessionStorage
        $a = new IlluminateSessionStorage();
        $b = new SessionStorage();

        $this->assertEquals($a, $b); //false

        $c = clone $b;
        $this->assertEquals($b, $c); //true

        $c->foo = 'bar';
        $this->assertEquals($b, $c); //false

        $d = new IlluminateSessionStorage();
        $this->assertEquals($a, $d); //true
```
Take a look at the very last line. It shows that `assertEquals` compare if the objects looks the same, but not if the variables reference the same object. 

Do as you suggested. (Use the original assert but with IlluminateSessionStorage.) It will be great.

Nyholm commented:
I think it looks very good. I was one minor issue with the test. Please fix it and squash your commits and I'll be happy to merge

Nyholm commented:
This will be a BC break for the users relying on the IlluminateStorage. 
Ping @andreascreten. Can we get your 50 cents on this PR. 

You will have to update your code like: 
```php
$ln = new LinkedIn(self::APP_ID, self::APP_SECRET, new IlluminateSessionStorage());
```

Or extend the LinkedIn class and make changes in the init function.

hannesvdvreken commented:
This is a BC break. You can modify it so IlluminateSessionStorage is still the default when nothing is injected through the constructor.

There is actually no reason to add an optional parameter to the constructor for constructor injection when there already is `Happyr\LinkedIn\LinkedIn::setStorage(DataStorageInterface $store)` available.

Otherwise you'll start adding additional optional arguments for the `request` and `urlGenerator` as well. This might start bloating the constructor and make it more complex for new users.

andreascreten commented:
@Nyholm, @hannesvdvreken is currently working on the project which uses this class, so he takes over from me :smile:

Nyholm commented:
Thank you @andreascreten 

Yes @hannesvdvreken It is a BC break. I can agree with you about the constructor argument. The problem with the current implementation is that people is installing the library without using the Facades correctly (see #22, #21, some of #20 and maybe #35).

Suggestion: Instead of creating the SessionStorage in `LinkedIn::init` maybe we should rewrite the `LinkedIn::getStorage` to something like: 
```php
public function getStorage()
{
    if ($this->storage == null) {
        $this->storage = new SessionStorage();
    }
    return $this->storage;
}
```

If you want to use the IlluminateSessionStorage you have to: 
```php
$ln = new LinkedIn(self::APP_ID, self::APP_SECRET);
$ln->setStorage(new IlluminateSessionStorage());
```

hannesvdvreken commented:
@Nyholm the problem is

```
if (class_exists(''...)) {
   ...
}
```

To detect being installed in a Laravel environment. You can't remove that because that will break backwards compatibity, but ideally one would keep the class as simple as possible and not allow framework specific code in there. Ideally one would create a ServiceProvider where a `LinkedIn` object is created an then injected with a `IlluminateSessionStorage` object before it is returned by the application container.

There is no way back now.

Nyholm commented:
> You can't remove that because that will break backwards compatibility

It will break BC, yes. But you are allowed to break backwards compatibility if (and only if)
* You bump the major version. 
* If the library has not reached its first stable version yet. 

So, yes. I can remove it. 

> Ideally one would create a ServiceProvider where a LinkedIn object is created an then injected with a IlluminateSessionStorage object.

Yes, exactly. See how I've done it with [Symfony](https://github.com/Happyr/LinkedIn-API-client/blob/master/docs/symfony.md)

hannesvdvreken commented:
:+1: version `0.5.0` it is then!

Nyholm commented:
@SmellyFish Can you make the changes according to the latest comments?

SmellyFish commented:
@Nyholm you're right. I didn't realize that `assertEquals()` was quite so loose. I moved the default session storage logic from `init()` to `getStorage()` like you had suggested. I opted for using the stricter assert same assertions in the constructor tests and looking at the object attributes directly instead of using the accessor. I also added a couple of tests for `getStorage()` to vet the new logic. Let me know if I missed anything. Once you're happy with the changes, I'll squash my commits into one.

Nyholm commented on src/Happyr/LinkedIn/LinkedIn.php:
```diff
@@ -90,32 +90,29 @@ class LinkedIn
+     * @param DataStorageInterface $storage [optional] the session storage object to use
      */
-    public function __construct($appId, $appSecret)
+    public function __construct($appId, $appSecret, DataStorageInterface $storage=null)
```
Remove the constructor param

Nyholm commented on src/Happyr/LinkedIn/LinkedIn.php:
```diff
@@ -90,32 +90,29 @@ class LinkedIn
         $this->appSecret = $appSecret;
 
-        $this->init();
+        $this->init($storage);
```
and this $storage parameter

Nyholm commented on src/Happyr/LinkedIn/LinkedIn.php:
```diff
@@ -90,32 +90,29 @@ class LinkedIn
-        } else {
-            $this->storage = new SessionStorage();
-        }
+        $this->storage = $storage;
```
And this

Nyholm commented on src/Happyr/LinkedIn/LinkedIn.php:
```diff
@@ -575,6 +572,9 @@ public function setStorage(DataStorageInterface $storage)
      */
     public function getStorage()
     {
+        if (null == $this->storage) {
```
Use strict comparison (===)

Nyholm commented:
Thank you @SmellyFish  for working on this. As @hannesvdvreken pointed out we should not bloat our constructor. So lets remove `DataStorageInterface` from the constructor completely.

SmellyFish commented:
No problem. Will update the code like you asked.

SmellyFish commented:
Would it make sense to get rid of the `init()` method? Seems like its logic could just be moved into the constructor.

Nyholm commented:
The purpose of the `LinkedIn::init` was to make it easier to extend the LinkedIn class and override the that function in order to use another URLGenerator och Request.

If you want to, you might remove it completely and instantiate the objects in their respectively get-method. Just like you did with the DataStorage.

Nyholm commented on Readme.md:
```diff
@@ -85,6 +85,64 @@ echo "<a href='$url'>Login with LinkedIn</a>";
 
+You can easily implement your own session storage device and use it with this library. Just make sure your session
+storage class implements the `\Happyr\LinkedIn\Storage\DataStorageInterface` interface (or write an adapter that
+uses your class and implements that interface as shown below).
```
Great that you documented this feature!
But you don't have to put the entire file in the readme, just make a link to the document.

SmellyFish commented:
@Nyholm It looks like the library hasn't been patched to comply with the [upcoming LinkedIn API update](https://developer.linkedin.com/support/developer-program-transition). I'll have to add more modifications to the code in the next week or two. I'm going to close this PR for now and re-open a new one once I've made all of the necessary updates.

Nyholm commented:
Okey, sure. It is highly appreciated. I will merge #39 now, make sure to pull those changes.
